src/vellum/workflows/context.py

```diff
@@ -0,0 +1,42 @@
+from contextlib import contextmanager
+import threading
+from typing import Any, Callable, Dict, Iterator, TypeVar, cast
+
+from vellum.workflows.events.types import ParentContext
+
+ExecutionLocalContext = Dict[str, Any]
+_CONTEXT_KEY = "_execution_context"
+
+local = threading.local()
+
+_context_lock = threading.Lock()
+
+F = TypeVar("F", bound=Callable[..., Any])
+T = TypeVar("T")
+P = TypeVar("P", bound=ParentContext)
+
+
+def get_execution_context() -> ExecutionLocalContext:
+    """Retrieve the current execution context."""
+    return getattr(local, _CONTEXT_KEY, {})
+
+
+def set_execution_context(context: ExecutionLocalContext) -> None:
+    """Set the current execution context."""
+    setattr(local, _CONTEXT_KEY, context)
+
+
+def get_parent_context() -> ParentContext:
+    return cast(ParentContext, get_execution_context().get("parent_context"))
+
+
+@contextmanager
+def execution_context(**kwargs: Any) -> Iterator[None]:
+    """Context manager for handling execution context."""
+    prev_context = get_execution_context()
+    new_context = {**prev_context, **kwargs}
```

Comment:
[nit] can we make the `kwargs` types explicit? ie, `parent_context: ParentContext`.

[nit] It's still odd to me that `new_context` has the same args as `prev_context`. I would've expected the new context to be constructed and point to prev context, but maybe I'm missing something

EDIT: Let's dismiss the second nit, I understand now that `new_context` and `prev_context` are part of the `ExecutionContext`, _not_ `ParentContext`

Reply:
I'll see about adding some typing to make it easier on the eyes!